test/jdk/java/lang/String/CompactString/MaxSizeUTF16String.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.junit.jupiter.params.provider.Arguments;
+import static org.junit.jupiter.api.Assertions.*;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+
+/*
+ * @test
+ * @bug 8077559 8321180
+ * @summary Tests Compact String for maximum size strings
+ * @requires os.maxMemory >= 7g & vm.bits == 64
+ * @run junit/othervm -XX:+CompactStrings -Xmx7g MaxSizeUTF16String
+ * @run junit/othervm -XX:-CompactStrings -Xmx7g MaxSizeUTF16String
+ */
+
+public class MaxSizeUTF16String {
+
+    private final static int MAX_UTF16_STRING_LENGTH = Integer.MAX_VALUE / 2;
+
+    private final static String EXPECTED_OOME_MESSAGE = "UTF16 String size is";
+
+    // Create a large UTF-8 byte array with a single non-latin1 character
+    private static byte[] generateUTF8Data(int byteSize) {
+        byte[] nonAscii = "\u0100".getBytes();
+        int nonAsciiSize = nonAscii.length;
+        int asciisize = byteSize - nonAsciiSize;
+        byte[] arr = new byte[asciisize + nonAsciiSize];
```

Comment:
```suggestion
        byte[] arr = new byte[byteSize];
```
Note that `asciisize + nonAsciiSize` = `byteSize - nonAsciiSize + nonAsciiSize` = `byteSize`, even in the presence of overflows. Then `asciisize` becomes useless.